File: binarySearch.cpp
#include<iostream>
#include<vector>
using namespace std;


int binarySearch( vector<int> inp, int target ){

    int s = 0, e = inp.size()-1 ,mid;

    
    while(s<=e){

        mid = (s + e)/2;

        if(inp[mid] == target){
            return mid;
            break;
        }
        else if(inp[mid] > target){
            e = mid-1;
        }
        else 
            s = mid+1;
    }
    return -1;
}

int main(){

    cout<<binarySearch( {1,2,3,7,8,9,10,11}, 10 )<<endl;

    return 0;
}

File: JAVA/BinarySearch.java
//package JAVA;

public class BinarySearch {
    
    public static void main( String []args){
        
        int []arr = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
        
        int position = binarySearch(9, arr);
        
        System.out.println(position);
    }

    public static int binarySearch( int target, int []inp ){

        int start = 0, end = inp.length - 1, mid;

        while(start <= end){
            
            //mid calculation is such so that adding 
            //start and end wont exceed the integer limit
            mid = start + (end - start)/2;

            if(inp[mid] == target) return mid;
            
            else if(inp[mid] < target) start = mid + 1;

            else end = mid - 1;
        
        }
        return -1;
    }

}
